### ISSUE TITLE
Opportunity to choose to insert rich or plain text from the copy buffer

### ISSUE THREAD

**Issue by joeUser555** [freeplane/freeplane#1050]
**The problem**
While editing a freeplane map ist is not visible if the edited node contains HTML text or plain text.
When transforming the freeplane xml file this evolves to a problem because a plain TEXT attribute can be handled completely different to the richcontext node.
This problem occured when I copied text from pycharm to freeplane.
It seems pycharm offers the copy content although in HTML not only text.

**Solution 1**
It would be nice to make visible if a node contains richcontent using differen text style or a frame on the node text.
**Solution 2**
Alternatively the user can be asked for the insert mode of the copy buffer content (plain or rich).
Or there could be different shortcuts for inserting the copy buffer.
CTRL-V always inserts plain text from the copy buffer
and 
SHIFT-CTRL-V inserts rich text from the copy buffer.
**Solution 3**
Better solution there could be a config option to determine whether copy content gets to inserted as rich or plain text.
**Solution 4**
Offer an action to convert rich content to plain text. '
Which may be reversible or not, because rich text information gets lost in this process
**Solution 5**
Offer an export action for saving the complete file as plain attribute xml file. Which should convert all rich text nodes into plain text before saving. 

**Conclusion**
Solution 2 seems to be the simplest and the easiest one to implement.

**Comment by macmarrum** [comment]
A good point.

**Solution 1**
It's possible to add a Map Conditional Style with a Script Filter
```groovy
node.text.startsWith('<html')
```
It would be best to create a new Style (e.g. "Rich Content") with an icon or some other indicator for your rich-content nodes, and use it with the Script Filter.

**Solution 4**
This can be achieved with a script, e.g.
```groovy
if (node.text.startsWith('<html') && !node.plainText.startsWith('='))
    node.text = node.plainText
```
which will convert selected nodes to plain text.

**Comment by EdoFro** [comment]
there is also a menu command to transform a node text to plain text. 
I’m not at my computer, but I think it jas the hot key alt+p

**Comment by macmarrum** [comment]
A good point, Edo.

There is `Format->Remove format`, which takes care of node-level formatting, like node shape, icon size and other formatting properties from `Tool panel`.

And to remove the formatting of text itself, i.e. when editing node core, there is `Ctrl+T`. In SimplyHTML it's `Format->Remove formatting`. I believe the text needs to be highlighted first, i.e. it works on a selection.

**Comment by joeUser555** [comment]
Thank you for the insights. 
That is a good work around so far. 

But I still think Solution 2 would be a good way for a future release, 
because if the user can choose(determine) the insert format while copying 
it can save him unnecessary work after copying.

**Comment by EdoFro** [comment]
Here is the way I use Freeplane to not have those problems:

### Preferences:

1. `Preferences…->Behaviour->Confirmations->Use rich text for pasted nodes` set to **No**

![image](https://user-images.githubusercontent.com/35700575/227200664-cd1bc18d-5ced-4136-8219-766a2d60e88e.png)

2. `Preferences…->Behaviour->Rich-Text Editor->Default paste mode`  set to **Paste as plain-text**

![image](https://user-images.githubusercontent.com/35700575/227202404-6f444e80-0196-4b12-a022-85f12cc28304.png)

### Considerations:

1. to paste as plain text just use "Paste" (Ctrl + V)
2. to paste as rich text or another special format use "Paste as ..." (Ctrl + Shift + V)

This works also when pasting inside the note editor (same for details and notes).


Hope it helps,

edo

**Comment by EdoFro** [comment]
Here is the way I use Freeplane to not have those problems:

### Preferences:

1. `Preferences…->Behaviour->Confirmations->Use rich text for pasted nodes` set to **No**

![image](https://user-images.githubusercontent.com/35700575/227200664-cd1bc18d-5ced-4136-8219-766a2d60e88e.png)

2. `Preferences…->Behaviour->Rich-Text Editor->Default paste mode`  set to **Paste as plain-text**

![image](https://user-images.githubusercontent.com/35700575/227202404-6f444e80-0196-4b12-a022-85f12cc28304.png)

### Considerations:

1. to paste as plain text just use "Paste" (Ctrl + V)
2. to paste as rich text or another special format use "Paste as ..." (Ctrl + Shift + V)

This works also when pasting inside the note editor (same for details and notes).


Hope it helps,

edo

**Comment by joeUser555** [comment]
@EdoFro
Thank you for this hint, that helps a lot.
This means Solution 3 is already implemented but I did not know.

**Comment by joeUser555** [comment]
But having different shortcuts for inserting rich or plain text on the fly could still be an option
considering the opportunity to temporarily override the options default. 
As I said earlier this is neither urgent nor important.
A future release would be sufficient.

**Comment by joeUser555** [comment]
If you are working on different map simultaneously it could be could be much more convenient to have the opportunity to choose insert format on the fly as you may like to insert rich content in one map an plain text in the other.
I guess the defaults in the options dialog are valid for all open map files .
It may be an other feature request to have such options on a per file basis to have different settings for different map files.

**Comment by joeUser555** [comment]
Since the storage format is XML based it should be easy to extend the XML schema with file edit options 
to implement file based edit options.

**Comment by dpolivaev** [comment]
I think you should learn to script so that you can implement all you need for yourself.

**Comment by macmarrum** [comment]
@joeUser555 for `Preferences…->Behaviour->Confirmations->Use rich text for pasted nodes` there is the option to "Ask" each time content is pasted

**Comment by joeUser555** [comment]
@dpolivaev 
That is a good idea. Is there a freeplane scripting tutorial out in the community ?

**Comment by joeUser555** [comment]
@macmarrum 
I have version 1.10.6 and this option is missing in my version.

![Bildschirmfoto vom 2023-03-23 14-13-48](https://user-images.githubusercontent.com/126019140/227215297-7bd61311-0125-4e83-80f4-c70a738dbb24.png)

![Bildschirmfoto vom 2023-03-23 14-15-47](https://user-images.githubusercontent.com/126019140/227215709-f5a23fae-ed7a-471d-ab9f-a15adc03c2c9.png)

**Comment by joeUser555** [comment]
@macmarrum 
By the way it is a difference  
to be able to choose insert format on the fly by choosing the shortcut 
or being anointed by the tool asking for this decision on every paste action

**Comment by macmarrum** [comment]
@joeUser555 

I understand your need. It's perfectly reasonable. Maybe if more people need it too, it will make sense to spend development effort and implement it in Freeplane core. Otherwise the scripting approach could be a better way. And it might bring you even more benefits, because you'd be able to create "macro commands" to automate your common tasks.

For Scripting documentation, please have a look at https://docs.freeplane.org/scripting/Scripting.html

The value "Ask" for `Preferences…->Behaviour->Confirmations->Use rich text for pasted nodes` is available also in v1.10.6, but it's tricky to find. It's because the option has 3 values Yes/No/Ask, which come one after another.

You can click your "Nein" once or twice to see it change to the other values. Click it three times to change back to "Nein".

**Comment by EdoFro** [comment]
Dear @joeUser555,
I recomend you to make [that changes in preferences](https://github.com/freeplane/freeplane/issues/1050#issuecomment-1481108227) and use Freeplane for a time, so you can refine the feature request.
I have learned how to script and I can say it could be a really easy script to do, but I have no need for it.

I have also learned to use rich content in my nodes only if it's strictly necessary and only if it adds value to the map. I define most formatting as styles in the map.
That way:
- I can paste content from any source and I don't have to worry about the styles and formating used in that webpages
- the map looks good, because colors, fonts, etc. are uniformed
- I have no errors because of bad or incomplete html content
- I can focus on content, not format.

jm2c,

edo

**Comment by joeUser555** [comment]
@macmarrum 
I found the option.
In my current situation it is sufficient to fix the option to "Nein/no".
The option "Ask" is rarely feasible because in most cases you want to insert rich or plain text 
and only in some rare cases you want the opposite behavior than the default.

Thanks for the link to the scripting documentation.

**Comment by joeUser555** [comment]
@EdoFro 
Yes I am using Freeplane with the default "no" which is ok for me for now.
Different shortcuts for explicit insert behavior is not necessary immediately.
Since Freeplane is implemented in Java I can take a look at the source code when I will have more time for that.
Perhaps I can volunteer to implement that feature later.

Please keep this issue open so I can come back to the discussion when I have time to get back to it's implementation.
